carcara/src/parser/mod.rs

```diff
@@ -407,6 +410,68 @@ impl<'a, R: BufRead> Parser<'a, R> {
                 SortError::assert_eq(&Sort::RegLan, sorts[1].as_sort().unwrap())?;
                 SortError::assert_eq(&Sort::String, sorts[2].as_sort().unwrap())?;
             }
+            Operator::BvNot | Operator::BvNeg => {
+                assert_num_args(&args, 1)?;
+                for s in sorts {
+                    let s = s.as_sort().unwrap().clone();
+                    if !matches!(s, Sort::BitVec(_)) {
+                        return Err(ParserError::ExpectedBvSort(s));
+                    }
+                }
+                // TODO: create args value assertion 
+                // assert_args_value(&args, 1..)?;
+            }
+            Operator::BvBbTerm => {
+                assert_num_args(&args, 1..)?;
+                for s in sorts {
+                    let s = s.as_sort().unwrap().clone();
+                    SortError::assert_eq(&Sort::Bool, &s)?;
+                }
+            }
+            Operator::BvConcat
+            | Operator::BvAdd
+            | Operator::BvMul
+            | Operator::BvAnd
+            | Operator::BvOr
+            | Operator::BvXor => {
+                assert_num_args(&args, 2..)?;
+                for s in sorts {
+                    let s = s.as_sort().unwrap().clone();
+                    if !matches!(s, Sort::BitVec(_)) {
+                        return Err(ParserError::ExpectedBvSort(s));
+                    }
+                }
```

Comment:
Here, besides making sure that the terms are all bitvectors, we also have to make sure they have the same size (e.g., `(bvadd #b01 #b001)` should be rejected because the arguments have different widths). I think a nice way to do this is to first check that the first argument is a bitvector, and then checking that all arguments have the same sort.
```suggestion
                assert_num_args(&args, 2..)?;
                let first_sort = sorts[0].as_sort().unwrap().clone();
                if !matches!(first_sort, Sort::BitVec(_)) {
                    return Err(ParserError::ExpectedBvSort(first_sort));
                }
                SortError::assert_all_eq(
                    &sorts
                        .iter()
                        .map(|op| op.as_sort().unwrap())
                        .collect::<Vec<&Sort>>(),
                )?;
```

Reply:
```rust
SortError::assert_all_eq(
                    &sorts
                        .iter()
                        .map(|op| op.as_sort().unwrap())
                        .collect::<Vec<&Sort>>(),
                )?;
```

is this code already enough to check if arguments have the same width?

Reply:
and also, I don't think all arguments should have the same width for `Concat`. (Double checking w/ you)

Reply:
one more thing: besides the check provided from the code above, do I also need to check the values of the values of the arguments for non-indexed operators?


something like `assert_args_value`

Reply:
> ```rust
> SortError::assert_all_eq(
>                     &sorts
>                         .iter()
>                         .map(|op| op.as_sort().unwrap())
>                         .collect::<Vec<&Sort>>(),
>                 )?;
> ```
> 
> is this code enough to check if arguments have the same width?

Yeah!

Reply:
> and also, I don't think all arguments should have the same width for `Concat`. (Double checking w/ you)

You're right, for string concatenation the arguments can have different widths (but they should all be bitvectors)

Reply:
> one more thing: besides the check provided from the code above, do I also need to check the values of the values of the arguments for non-indexed operators?
> 
> something like `assert_args_value`

No. You only have to check that for indexed operators, since they have `op_args` that need to follow certain restrictions.